### PR TITLE
Add pose graph edge visualization

### PR DESCRIPTION
**Pull request by hidmic**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (only connected to #407) |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | CPR's Jackal gazebo simulation |

---

## Description of contribution in a few bullet points

* Extended graph visualization to include edges, as shown below:

https://user-images.githubusercontent.com/13500507/155807874-45b8a2a5-23cc-4e15-b8bc-be76c0542ff4.mp4

## Description of documentation updates required from your changes

* None, as this piggybacks on the existing interactive mode.

### CODE REVIEW

File: slam_toolbox/src/loop_closure_assistant.cpp [hunk 1]
```diff
@@ -125,48 +125,93 @@ void LoopClosureAssistant::publishGraph()
...
-  std::unordered_map<int, Eigen::Vector3d>* graph = solver_->getGraph();
+  karto::MapperGraph * graph = mapper_->GetGraph();
 
-  if (graph->size() == 0)
+  if (!graph || graph->GetVertices().empty())
   {
     return;
   }
 
-  ROS_DEBUG("Graph size: %i",(int)graph->size());
```
Comment (SteveMacenski) [inline slam_toolbox/src/loop_closure_assistant.cpp:135]: Please readd the debug statement

File: slam_toolbox/src/loop_closure_assistant.cpp [hunk 2]
```diff
@@ -125,48 +125,93 @@ void LoopClosureAssistant::publishGraph()
...
+  // schedule all existing markers for removal
+  for (auto it = markers_.begin(); it != markers_.end(); ++it) {
+    it->second.action = visualization_msgs::Marker::DELETE;
+  }
+
+  visualization_msgs::Marker vertex_marker =
+    vis_utils::toVertexMarker(map_frame_, "slam_toolbox", 0.1);
 
-  for (ConstGraphIterator it = graph->begin(); it != graph->end(); ++it)
+  for (const auto & vertices_per_sensor : graph->GetVertices())
```
Comment (SteveMacenski) [inline slam_toolbox/src/loop_closure_assistant.cpp]: Please use the old style of for loops to be consistent in codebase

File: slam_toolbox/src/loop_closure_assistant.cpp [hunk 3]
```diff
@@ -125,48 +125,93 @@ void LoopClosureAssistant::publishGraph()
...
+      vis_utils::toEdgeMarker(map_frame_, "slam_toolbox", 0.05);
+
+    for (auto * edge : graph->GetEdges()) {
+      karto::LocalizedRangeScan * source_scan = edge->GetSource()->GetObject();
+      karto::LocalizedRangeScan * target_scan = edge->GetTarget()->GetObject();
+      const karto::Pose2 source_pose = source_scan->GetCorrectedPose();
+      const karto::Pose2 target_pose = target_scan->GetCorrectedPose();
+
+      edge_marker.id =
+          (source_scan->GetUniqueId() << 16) + target_scan->GetUniqueId();
```
Comment (SteveMacenski) [inline slam_toolbox/src/loop_closure_assistant.cpp]: what's what this naming?

File: slam_toolbox/src/loop_closure_assistant.cpp [hunk 4]
```diff
@@ -125,48 +125,93 @@ void LoopClosureAssistant::publishGraph()
...
+  for (const auto & vertices_per_sensor : graph->GetVertices())
   {
-    m.id = it->first + 1;
-    m.pose.position.x = it->second(0);
-    m.pose.position.y = it->second(1);
+    for (const auto & id_and_vertex : vertices_per_sensor.second)
+    {
+      karto::LocalizedRangeScan * scan = id_and_vertex.second->GetObject();
+
+      vertex_marker.id = scan->GetUniqueId() + 1;
```
Comment (SteveMacenski) [inline slam_toolbox/src/loop_closure_assistant.cpp]: +1? Its been a bit since I've worked in this part of the code, why this?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: slam_toolbox/src/loop_closure_assistant.cpp (hunk 4 above)
Comment (hidmic) [reply slam_toolbox/src/loop_closure_assistant.cpp]: That's preexisting code. Commit history doesn't shed much light. We can remove it.

File: slam_toolbox/src/loop_closure_assistant.cpp (hunk 3 above)
Comment (hidmic) [reply slam_toolbox/src/loop_closure_assistant.cpp]: That's a best effort, arguably not elegant approach to deriving 32 bit edge IDs from 32 bit source and target vertices IDs. As long as vertices' unique IDs stay under 2^15 - 1 (IDs are signed for some reason), there won't be collisions. 

A hash function for a pair of integers may work too. Can't make guarantees about collision rates then.

File: slam_toolbox/src/loop_closure_assistant.cpp (hunk 4 above)
Comment (SteveMacenski) [reply slam_toolbox/src/loop_closure_assistant.cpp]: Lets keep it if its preexisting, that's fine.

File: slam_toolbox/src/loop_closure_assistant.cpp (hunk 3 above)
Comment (SteveMacenski) [reply slam_toolbox/src/loop_closure_assistant.cpp]: Couldn't you make vertices count from 1,2,3,4... while you make edges count down 2^15, 2^15-1, ...? Might make things easier than a hashing function

File: slam_toolbox/src/loop_closure_assistant.cpp (hunk 1 above)
Comment (hidmic) [reply slam_toolbox/src/loop_closure_assistant.cpp:135]: Done in https://github.com/SteveMacenski/slam_toolbox/pull/467/commits/9f4eb994925ce5246694f4d76656802bba562ef5.

File: slam_toolbox/src/loop_closure_assistant.cpp (hunk 2 above)
Comment (hidmic) [reply slam_toolbox/src/loop_closure_assistant.cpp]: Done in https://github.com/SteveMacenski/slam_toolbox/pull/467/commits/9f4eb994925ce5246694f4d76656802bba562ef5.

File: slam_toolbox/src/loop_closure_assistant.cpp (hunk 3 above)
Comment (hidmic) [reply slam_toolbox/src/loop_closure_assistant.cpp]: Do you mean attaching a unique ID to edges? Those will collide with vertex IDs over time too. 

Issue here is that we need some function that relates graph vertices and edges with their markers if we want to remove (resp. update) the latter when the former are removed (resp. updated).

--- 

Edit: hmm, but we don't need to use IDs for that. We could have a collection of `(vertex, marker)` tuples and a collection of `(edge, marker)` tuples (maybe indexed to speed up lookups). Then we can have a separate counter for marker IDs that just rolls over. WDYT?

Comment (hidmic) [reply slam_toolbox/src/loop_closure_assistant.cpp]: @SteveMacenski PTAL at https://github.com/SteveMacenski/slam_toolbox/pull/467/commits/07961ed95217360e067a4505414077a1d4533890

**Review by SteveMacenski — APPROVED** [review]
LGTM, you ready for this to go in? Can you show me an rviz screen shot of this working (mostly just for historical records)

**Comment by hidmic** [comment]
@SteveMacenski I added a short screencast of RViz to this PR's description showcasing the feature . PTAL!

**Comment by SteveMacenski** [comment]
If you're open to it, I'd be happy to merge a forward port to ROS 2

**Pull request merged**